x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/fields/ExtractedField.java

```diff
@@ -88,12 +99,86 @@ public static ExtractedField newField(String alias, String name, ExtractionMetho
+
+        GeoField(String alias, String name, ExtractionMethod extractionMethod) {
+            super(alias, name, extractionMethod);
+            internalExtractor = extractionMethod.equals(ExtractionMethod.SOURCE) ?
+                new FromSource(alias, name, extractionMethod) :
+                new FromFields(alias, name, extractionMethod);
+        }
+
+        @Override
+        public Object[] value(SearchHit hit) {
+            Object[] value = internalExtractor.value(hit);
+            if (value.length == 2 && internalExtractor.getExtractionMethod().equals(ExtractionMethod.SOURCE)) { // geo_point as array
+                return new Object[] {value[0] + "," + value[1]};
+            }
+            if (value.length != 1) {
+                throw new IllegalStateException("Unexpected value count for a geo point field: " + value);
+            }
+            if (value[0] instanceof String) {
+                value[0] = handleString((String) value[0]);
+            } else if(value[0] instanceof Map<?, ?>) {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> geoObject = (Map<String, Object>) value[0];
+                value[0] = handleObject(geoObject);
+            } else if(value[0] instanceof List<?>) {
+                @SuppressWarnings("unchecked")
+                List<Double> coordinates = (List<Double>) value[0];
+                assert coordinates.size() == 2;
+                value[0] = coordinates.get(0) + "," + coordinates.get(1);
+            } else {
+                throw new IllegalStateException("Unexpected value type for a geo point field: " + value[0].getClass());
+            }
+            return value;
+        }
+
+        private String handleString(String geoString) {
+            try {
+                if (geoString.startsWith("POINT")) { // Entry is of the form "POINT (-77.03653 38.897676)"
+                    Geometry geometry = wkt.fromWKT(geoString);
+                    if (geometry.type() != ShapeType.POINT) {
+                        throw new IllegalArgumentException("Unexpected non-point geo type: " + geometry.type().name());
+                    }
+                    Point pt = ((Point)geometry);
+                    return pt.getLat() + "," + pt.getLon();
+                } else if (geoString.contains(",")) { // Entry is of the form "38.897676, -77.03653"
+                    return geoString.replace(" ", "");
+                } else { // This may be a geohash, attempt to decode
+                    Point pt = Geohash.toPoint(geoString);
+                    return pt.getLat() + "," + pt.getLon();
+                }
+            } catch (IOException | ParseException ex) {
+                throw new IllegalArgumentException("Unexpected value for a geo field: " + geoString);
+            }
+        }
+
+        private String handleObject(Map<String, Object> geoObject) {
+            if ("point".equals(geoObject.get("type"))) { // geo_shape
```

Comment:
is this case sensitive?

Reply:
@dimitris-athanasiou , good question. I will dig into that.